pennylane/gradients/parameter_shift.py

```diff
@@ -174,19 +176,14 @@ def _single_meas_grad(result, coeffs, unshifted_coeff, r0):
 def _multi_meas_grad(res, coeffs, r0, unshifted_coeff, num_measurements):
     """Compute the gradient for multiple measurements by taking the linear combination of
     the coefficients and each measurement result."""
-    g = []
     if r0 is None:
         r0 = [None] * num_measurements
-    for meas_idx in range(num_measurements):
-        # Gather the measurement results
-        meas_result = [param_result[meas_idx] for param_result in res]
-        g_component = _single_meas_grad(meas_result, coeffs, unshifted_coeff, r0[meas_idx])
-        g.append(g_component)
+    if res == ():
+        res = tuple(() for _ in range(num_measurements))
+    return tuple(_single_meas_grad(r, coeffs, unshifted_coeff, r0_) for r, r0_ in zip(res, r0))
 
-    return tuple(g)
 
-
-def _evaluate_gradient(tape, res, data, r0):
+def _evaluate_gradient(tape_specs, res, data, r0, batch_size):

```

Comment:
Effective changes for this function:
- Use precomputed `tape_specs` instead of the full tape.
- add `batch_size` as input because we need that information now.
- Add support for broadcasting for multiple measurement and/or shot vectors (both previously not supported)
- Cover 8 scenarios now: (single/multiple measurements) x (single shots/shot vector) x (broadcast=True/False); previously only covered 4: (single/multiple measurements) x (single shots/shot vector), and broadcasting worked "accidentally" in the only supported case (single measurement + single shots).
- `_multi_meas_grad` changed expected axes, adapt to that change.